### PR TITLE
Only keep existing mutation timestamp when user is replication

### PR DESCRIPTION
**Pull request by benbarclay**

### CODE REVIEW

File: pkg/replication/admission.go [hunk 1]
```diff
@@ -37,6 +38,9 @@ import (
 )
 
 const (
+	// User account this component uses
```
Comment (nilebox) [inline pkg/replication/admission.go]: "this component"? replication admission is a separate component from replication controller...
I would rephrase comment to explicitly mention replication controller.

File: pkg/replication/admission.go [hunk 2]
```diff
@@ -332,8 +336,12 @@ func (ac *AdmissionContext) validateLocationsAndTransforms(sd *comp_v1.ServiceDe
 	return RejectionMessage(strings.Join(rejectionMessages, ", "))
 }
 
-func createAnnotationPatch(operation admissionv1beta1.Operation, newSD *comp_v1.ServiceDescriptor) ([]byte, error) {
-	ts, err := generateUpdated(operation, newSD)
+func isReplication(userInfo authenticationv1.UserInfo) bool {
```
Comment (nilebox) [inline pkg/replication/admission.go]: `isReplicationController`?

File: pkg/replication/admission.go [hunk 3]
```diff
@@ -489,15 +497,14 @@ func addAnnotation(key, value string) util.Patch {
 	}
 }
 
-func generateUpdated(operation admissionv1beta1.Operation, newSD *comp_v1.ServiceDescriptor) (time.Time, error) {
-	if operation != admissionv1beta1.Update {
+func generateUpdated(isReplication bool, operation admissionv1beta1.Operation, newSD *comp_v1.ServiceDescriptor) (time.Time, error) {
+	if operation != admissionv1beta1.Update || !isReplication {
```
Comment (nilebox) [inline pkg/replication/admission.go]: Hmm... So this code is written in a way that if replication controller doesn't set the timestamp (why? is that possible?), we will generate a timestamp.

Maybe instead we could always parseTimestamp for replication controller and return error if annotation is missing?

File: pkg/replication/admission.go (hunk 1 above)
Comment (benbarclay) [reply pkg/replication/admission.go]: I view replication as an entire subsystem. The implementation is split across a controller and a couple of webhooks. I take it you see it differently?

File: pkg/replication/admission.go (hunk 3 above)
Comment (benbarclay) [reply pkg/replication/admission.go]: afaict the only situation we *don't* want to generate a timestamp is if the replication controller is doing an update on an existing thing. In all other cases we do want to generate a timestamp. Isn't that what's going on here?

Comment (nilebox) [reply pkg/replication/admission.go]: Yes, but what is the case when replication controller doesn't provide a timestamp? I guess to rephrase the requirement: replication controller _must_ set timestamp and we don't touch it (but validate it's set). In all other cases we generate timestamp.
This way it will be easier to catch bugs in replication controller I think

File: pkg/replication/admission.go (hunk 1 above)
Comment (nilebox) [reply pkg/replication/admission.go]: Fair enough, but in this case it's replication controller request to API server being intercepted by a webhook, so it seems useful to be more clear about what we do.

File: pkg/replication/admission.go (hunk 3 above)
Comment (benbarclay) [reply pkg/replication/admission.go]: Fair call, I'll fix that up

Comment (nilebox) [reply pkg/replication/admission.go]: Also not sure if we actually still need to check for `operation != admissionv1beta1.Update`? I find this code hard to understand. My idea:
```
if isReplication {
  ts, exists := newSD.Annotations[updatedKey]
  if !exists {
    return nil, errors.New("replication controller must set the mutation timestamp")  
  }
  return parseTimestamp(ts)
}

return time.Now(), nil
```

File: pkg/replication/admission.go [hunk 4]
```diff
@@ -489,16 +497,15 @@ func addAnnotation(key, value string) util.Patch {
...
+	if isReplication {
+		ts, exists := newSD.Annotations[updatedKey]
+		if !exists {
+			return time.Now(), errors.New("replication controller must set the mutation timestamp")
+		}
+		return parseTimestamp(ts)
 	}
-
-	return parseTimestamp(ts)
+	return time.Now(), nil
```
Comment (nilebox) [inline pkg/replication/admission.go]: Sorry, I think we may need to keep the timestamp if hash hasn't changed, e.g. when someone adds/deletes annotation or finalizer.
So you need to pass a flag if hash has changed?
